server/datastore/mysql/sessions.go

```diff
@@ -3,12 +3,80 @@ package mysql
 import (
 	"context"
 	"database/sql"
+	"errors"
+	"time"
 
+	"github.com/fleetdm/fleet/v4/server"
 	"github.com/fleetdm/fleet/v4/server/contexts/ctxerr"
 	"github.com/fleetdm/fleet/v4/server/fleet"
 	"github.com/jmoiron/sqlx"
 )
 
+const mfaLinkTTL = time.Minute * 15
+const mfaTokenEntropyInBytes = 32
+
+func (ds *Datastore) SessionByMFAToken(ctx context.Context, token string, sessionKeySize int) (*fleet.Session, *fleet.User, error) {
+	var userID uint
+	err := sqlx.GetContext(
+		ctx,
+		ds.reader(ctx),
+		userID,
+		"SELECT user_id FROM verification_tokens WHERE token = ? AND created_at >= NOW() - INTERVAL ? SECOND",
+		token,
+		mfaLinkTTL.Seconds(),
+	)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, nil, ctxerr.Wrap(ctx, notFound("Verification Token"))
+		}
+		return nil, nil, err
+	}
+
+	user, err := ds.UserByID(ctx, userID)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var session *fleet.Session
+	err = ds.withTx(ctx, func(tx sqlx.ExtContext) error {
+		if session, err = ds.makeSessionInTransaction(ctx, tx, user.ID, sessionKeySize); err != nil {
+			return err
+		}
+
+		// only delete token once we've successfully consumed it
+		if _, err = tx.ExecContext(ctx, "DELETE FROM verification_tokens WHERE token = ?", token); err != nil {
+			return err
+		}
```

Comment:
We need a cron anyways?

Reply:
Erring on the side of failing closed here. Treating these tokens as basically one time passwords, and if the network is flakey enough to have issues when redeeming the token in a quick API call the end user has bigger problems.

If we run into significant issues once this is out to customers we can re-evaluate, but one big reason I want these tokens invalidated on initial use is that they're going to show up in server logs as soon as the request comes in so (while that validity window is short) tailing server logs defeats MFA (as does a replay attack). And we have to put the token in the URL because I don't trust mail protection systems to pass through hash parameters on URLs to the browser from an email, so we can't just keep the token client-side.

Reply:
Re: needing a cron anyway, on one hand I'd normally say yes (to purge expired unused magic links), but given the low volume of those calls, the fact that a magic link was issued (which requires correct username and password) and not used is potentially of great interest for IR. So until we have more detail in the verification emails (which was out of scope for this iteration) I think we intentionally _don't_ do cleanup. Though if we had the magic link _prior_ to MFA we absolutely would need a cleanup cron.

Reply:
Sounds good.

> they're going to show up in server logs as soon as the request comes in so

Where exactly? Do you mean ingress/nginx logs?

Reply:
Correct.